src/backend/gl/src/native.rs

```diff
@@ -38,6 +43,79 @@ impl Fence {
     }
 }
 
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub(crate) enum BindingTypes {
+    UniformBuffers,
+    StorageBuffers,
+    StorageImages,
+    SampledImages,
+    SeparateImages,
+    SeparateSamplers,
+    SubpassInputs,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct DescRemapData {
+    bindings: HashMap<BindingTypes, HashMap<SetID, HashMap<BindingID, BindingID>>>,
+    names: HashMap<String, (BindingTypes, SetID, BindingID)>,
+    next_binding: BindingID,
+}
+
+impl DescRemapData {
+    pub(crate) fn new() -> Self {
+        DescRemapData {
+            bindings: HashMap::new(),
+            names: HashMap::new(),
+            next_binding: 0,
+        }
+    }
+
+    pub(crate) fn get_binding_insert(&mut self, btype: BindingTypes, set: SetID, binding: BindingID) -> BindingID {
+        if !self.bindings.contains_key(&btype) {
+            self.bindings.insert(btype, HashMap::new());
+        }
+
+        let bindings = self.bindings.get_mut(&btype).unwrap();
+        if !bindings.contains_key(&set) {
+            bindings.insert(set, HashMap::new());
+        }
+
+        let set_map = bindings.get_mut(&set).unwrap();
+        if !set_map.contains_key(&binding) {
+            set_map.insert(binding, self.next_binding);
+            self.next_binding += 1;
+        }
+
+        *set_map.get(&binding).unwrap()
+    }
+
+    pub(crate) fn get_binding(&self, btype: BindingTypes, set: SetID, binding: BindingID) -> Option<&BindingID> {
+        if !self.bindings.contains_key(&btype) {
+            return None
+        }
+
+        let bindings = self.bindings.get(&btype).unwrap();
```

Comment:
similarly, instead of checking for `contains_key` followed to unwrapping we could just chain all the things:
```rust
self.bindings
  .get(&btype)?
  .get(&set)?
  .get(&binding)
```